src/parser/jobs/mnk/BuffUptime.js

```diff
@@ -0,0 +1,161 @@
+import React, {Fragment} from 'react'
+
+import {ActionLink} from 'components/ui/DbLink'
+import ACTIONS, {getAction} from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {Rule, Requirement} from 'parser/core/modules/Checklist'
+import {Suggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+
+const GCD_CYCLE_LENGTH = 5
+
+export default class BuffUptime extends Module {
+	static handle = 'BuffUptime'
+	static dependencies = [
+		'checklist',
+		'combatants',
+		'enemies',
+		'invuln',
+		'suggestions',
+	]
+
+	// Dragon kick can be applied to multiple enemies
+	_lastDragonKickUse = {}
+	_lastTwinSnakesUse = null
+
+	_earlyDragonKicks = 0
+	_earlyTwinSnakes = 0
+
+	_gcdsSinceDK = {}
+	_gcdsSinceTS = 0
+
+	constructor(...args) {
+		super(...args)
+
+		// Hook all GCDs so we can count GCDs in buff windows
+		this.addHook('cast', {by: 'player'}, this._onCast)
+
+		// Apparently DK is a bit busted but let's use the debuff for now
+		this.addHook(['applybuff', 'refreshbuff'], {by: 'player', abilityId: STATUSES.BLUNT_RESISTANCE_DOWN.id}, this._onDragonKick)
+		this.addHook(['applybuff', 'refreshbuff'], {by: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this._onTwinSnakes)
+
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const action = getAction(event.ability.guid)
+
+		if (!action) {
+			return
+		}
+
+		if (action.onGcd) {
+			this._gcdsSinceDK[event.targetID] += 1
```

Comment:
Won't this result in nulls during the opener/when adds show up? `_gcdsSinceDK[event.targetID]` is only initialized in `_onDragonKick()`, but the standard opener has 3 GCDs before your first DK. Also for accuracy, shouldn't it increment all the targets you're tracking on each GCD cast, regardless of the target you actually hit?

Reply:
Except for Shinryu's Tail, I can't think of any time DK applies to more than 1 target on refresh. It's a debuff on the target so they get refreshed independently. The uptime calculation is also kinda wonky depending on how many targets. I'm open to ideas on how to work this, how do you handle Slashing on multiple targets for NIN?

I think to deal with initialisation/opener, we can check if it's null, or maybe if DK is on the target and it dropped. Thoughts? Thanks for the feedback!